### ISSUE TITLE
Proof AP11ROPE octals against Luminary 99

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#1041]
Just to be *absolutely sure* they're the same. I'm almost considering doing a fresh octal transcription just to eliminate any potential cross pollination -- but that may be a bit overly paranoid.

**Comment by rburkey2005** [comment]
I think it's an absolute necessity.

**Comment by thewonderidiot** [comment]
@rburkey2005 I was just going to start fiddling with octopus settings for the AP11ROPE octals, and it occurred to me that one quick check that AP11ROPE is the same as Luminary 99 R1 would be to look at the versions used for all of the subroutines. And it turns out, there's an interesting difference there.
![](http://www.ibiblio.org/apollo/ScansForConversion/AP11ROPE/0089.jpg)

AP11ROPE instantiates revision 001 of "AP11AID", while Luminary 99 R1 instantiates revision 001 of "LNYAIDE":
![](http://www.ibiblio.org/apollo/ScansForConversion/Luminary099/0089.jpg)

So, we should take an especially close look at everything in AP11AID.

**Comment by thewonderidiot** [comment]
Thinking on it a bit more, it's interesting that Luminary 99 R1 has "LNYAIDE" at all. Every other version of Luminary instead uses the LEMONAID subroutine. My best guess is that release "branches" like these had to "fork" subroutines to make their patches, so that mainline Luminary wouldn't be affected. (Several of Zerlina's sections have Z's in them -- presumably wherever different from Luminary. It uses ZERLAID instead of LEMONAID).

I know it's too late for this, but it would be very interesting to see which subroutine Allan Klumpp's Luminary 99 used. It's also very interesting that LNYAIDE is at revision 001. I wonder if subroutines start at 000 or 001... if the former, then I wonder if it's possible to rev them without revving Luminary. I.e., I wonder if the extra revision that Jim Kernan is thinking of happened to LNYAIDE without affecting the revision of Luminary proper...

**Comment by thewonderidiot** [comment]
Yeah, it is possible. Listed in the YUL programmer's digest:
```LYNN  ASSEMBLE NEW AGC4 SUBROUTINE INTPRET BY MUNTZ```
```NAN  ASSEMBLE REVISION 112 OF BLK2 SUBROUTINE LEADON BY MACBETH```

**Comment by thewonderidiot** [comment]
The main trouble with this theory, of course, is that CONTROLLED CONSTANTS is part of the base program, not contained in any of the subroutines... so LNYAIDE would have had to have been created for some other change, other than the constants on pages 48 and 49.

**Comment by rburkey2005** [comment]
Well, we'll know soon enough, I expect.  Unfortunately, we could easily have a situation like SUNBURST 37 being identical to SHEPATIN 0, and find out that Don merely was snapshotting this prior to making changes that he may or may not have gone ahead to make later.
At some point I'll ask Don about this difference, if you don't beat me to the punch, but my personal inclination is to get a fuller picture before doing so.

**Comment by thewonderidiot** [comment]
Okay, so I've done some more investigating into this difference, and the result is.... weird. Really weird. @indy91, you'll be interested in this too.

Disclaimer: throughout this comment I'm going to pretend we don't have Jim Kernan's recollections. This will be focusing exclusively on what we know from Luminary memos and listings.

I want to talk about Luminary 116 first. Luminary 116 was flown as a revision of mainline Luminary, instead of a small branch like 69R2 or 99R1. Partly thanks to this, [Luminary memo 107](http://www.ibiblio.org/apollo/Documents/LUM107_text.pdf) provides a very detailed description of what changed between 115 and 116.
```
A program error was corrected in the coding of R12 that had been
inserted for the implementation of PCR 839 (R12 and LR Re-position
Routine Improvements). The bug caused incorrect program per-
formance in the event that a PROCEED response was keyed in to the
flashing V05N09 (R1 = 523) display indicating failure of the LR
antenna reposition.
```

R12 is in the SERVICER log section, so it's safe to assume that the changed code was contained (almost) entirely in that section. Since this code is in a subroutine, however, the subroutine version call would have needed to have changed as well. According to ASSEMBLY AND OPERATION INFORMATION, SERVICER is part of the FLY subroutine. So looking at the SUBROUTINE CALLS section of the main program:
![lmy116_subros](https://user-images.githubusercontent.com/94056/28809334-abfb018e-7637-11e7-93b9-7c2ecec421f9.png)

FLY is unique in that it has an asterisk at the start of its line. So perhaps that indicates change a change from the previous revision! Going into the SERVICER and searching for "alarm" leads to this little chunk of code:
![lmy116_changes](https://user-images.githubusercontent.com/94056/28809429-2d5bfe72-7638-11e7-819e-f5e2f07e6efe.png)

Surrounding code confirms this has to do with radar positioning. If PROCEED is pressed, the code jumps to P1CHK. And sure enough, 4 lines here have asterisks at the start -- so the behavior on PROCEED was changed, exactly as Luminary Memo 107 said. So I'm pretty sure it's safe to say that GAP prefixes lines changed from the previous revision with an asterisk.

Okay, jumping back to LMY99 R1. Here are the ephemeris constants:
![lmy99r1_ephem1](https://user-images.githubusercontent.com/94056/28809577-1785ba38-7639-11e7-8e8c-48ae8572ca2a.png)
![lmy99r1_ephem2](https://user-images.githubusercontent.com/94056/28809578-19d7de10-7639-11e7-9bb8-45167c702eb7.png)

No asterisks! Here's the SUBROUTINE CALLS table:
![lmy99r1_subros](https://user-images.githubusercontent.com/94056/28809613-3e86750a-7639-11e7-9b30-e81b87349368.png)

So, LNYAIDE, which we've been talking about as suspicious above, had a version change from LMY99 R0. Assuming that is identical to Luminary revision 99, then, we "know" the change is somewhere in bank 5. So I started looking for any code that is in LEMONAID/LNYAIDE that is _also_ in bank 5. It turns out there's only two things -- the downlink lists plus some associated constants, and parts of FRESH START AND RESTART. A quick scan through the downlink lists unsurprisingly shows no asterisks.

FRESH START AND RESTART, however, does have surprises. STARTSB1 is one of the routines in the bank 5 part of FRESH START, and it has two asterisks:
![lmy99r1_startsb1](https://user-images.githubusercontent.com/94056/28809838-bd4383a0-763a-11e7-8402-1bd6beae1968.png)

So, if the assumptions above are correct, the lines `CAF POSMAX` and `CAF EBANK6` are what changed from Luminary 99 to Luminary 99R1. Which is reeeaaally weird. AP11ROPE has asterisks on exactly these same two lines, and an asterisk on its AP11AID subroutine call. Neither have any sort of comments in this area.

I think if we ask Don about AP11AID, we should ask explicitly about these two lines to see if anything shakes loose in his head. I'm going to spend more time searching for anywhere where we might be able to figure out what each of the LNY- anomalies fixed in Luminary 100 are.

Ron, you may have sent it to me before, but I seem to have lost it -- what was the banksum for Allan Klumpp's listing?

**Comment by indy91** [comment]
Looking at Luminary069, the change to FRESH START was that the beginning of STARTSB1 was moved from line 0394 to 0388, so from "CAF EBANK6" to "CAF POSMAX". That woud be consistent with the two asterisks. And it's the only change in that section from Luminary069 to 099.

**Comment by thewonderidiot** [comment]
Oh, good catch!! I totally missed that. So if, hypothetically, that was the difference between Luminary 99 and LMY99 R1, then I would expect Allan's banksum to be **44523**.

**Comment by rburkey2005** [comment]
According to Allan, bank 6 was 44523.

**Comment by rburkey2005** [comment]
The observation about the asterisks is great.  I never had any idea.

**Comment by thewonderidiot** [comment]
Awesome! Perfect match. :smile:

In that case, it looks like Luminary 99R1 was an emergency restart-related anomaly patch for Luminary 99, rather than an ephemeris constant update. Jim must have been remembering something else.

@rburkey2005, we have enough to exactly reconstruct revision 99 of Luminary with this. How do you think we should handle naming? What's been in the Luminary099 folder all of these years is actually LMY99 rev 1... so I'd almost be inclined to rename it, but that could get confusing real fast.

**Comment by thewonderidiot** [comment]
And yeah, I didn't either. I probably wouldn't have noticed if the SUBRO call lines weren't so empty. I wonder if we could use them to reconstruct any other versions... though this is the only case in which we'd have a banksum to verify against. At the very least we could put in comments pointing out the changed code.

This also makes me really curious about the differences between 209 and 210 -- I'm going to have to investigate that. The SUBROUTINE CALLS page has an asterisk next to *every single* SUBRO call. So either they went through and did a big comment text cleanup, or there were a lot more changes between the two than I thought.

**Comment by rburkey2005** [comment]
@thewonderidiot I would suggest calling a reconstructed Luminary 99 "LUM99R0", since I called the reconstructed R2 "LUM99R2".

**Comment by thewonderidiot** [comment]
Alright, I've added it as "LMY99R0", since that's a teensy bit more accurate. Funnily enough, it turns out there's one more asterisk in LMY99 Rev 1, and we've known about it all along.
![l99_sections](https://user-images.githubusercontent.com/94056/28859255-2816b898-770b-11e7-8486-4e192af48be4.png)

That asterisk was transcribed and (understandibly) has survived proofing. But it's one character too far to the left to be part of the comment. :smile: I guess they figured since it was a release program, they should go through the trouble of updating ASSEMBLY AND OPERATION INFORMATION to show the correct subroutine name, if they were going to change it.

As it turns out, Don did _not_ bother to make this change when putting together AP11ROPE.
![ap11_sections](https://user-images.githubusercontent.com/94056/28859299-65db8b22-770b-11e7-8ce7-0ce507a94acc.png)

I wonder how many more of our comments contain the "line updated" asterisk...

**Comment by rburkey2005** [comment]
I certainly mistakenly included a number of the asterisks myself ... and removed a number of them when transcribing/proofing SUNBURST 37, but I know I didn't catch all of them even in SUNBURST 37.  We probably ought to grep them all and keep a list, because now that we know what they are, they're a resource that will be difficult to replace if/when somebody goes through the source files and corrects them all by removing all the asterisks.

**Comment by thewonderidiot** [comment]
Yeah. I'll add some comments to these lines in Luminary 99R1 and Luminary 116. (I already did so for LMY99R0). At some point I'll go through and try to seek all of them out.

Getting back on this issue's topic, I've created and pushed an initial binsource for AP11ROPE. I used the stock training data for Tesseract, and the results seem pretty darn good -- seven banks have immediately correct checksums, and the remaining checksums look generally pretty close, so there is probably only a couple of errors per bank remaining. I haven't done any proofing yet. Here octopus-prepared images I used to make the binsource: [ap11rope_prepared.zip](https://github.com/virtualagc/virtualagc/files/1203497/ap11rope_prepared.zip)

**Comment by rburkey2005** [comment]
Zerlina is going so slowly that it looks like it will be some time before I can get to it. Had the best a diff against R1? That might be the quickest way to check it.


Sent on my Virgin Mobile Samsung Galaxy S® 5

-------- Original message --------

**Comment by thewonderidiot** [comment]
I was just gonna knock this out while you're working through Zerlina. I did a quick diff between Luminary 99R1's and AP11ROPE's oct2bin.proof files and there are indeed only a couple of differences per bank. It should go by very quickly.

**Comment by rburkey2005** [comment]
@thewonderidiot Per our emails:  I've gone ahead and fixed problems in the octals by checking diffs vs Luminary99R1 (yesterday), and I've completely proofed using octopus/ProoferBox.py (today), so I view the described task as having been completed.  But since you had assigned yourself this and have added lots of comments not directly related to proofing the octals, I was reluctant to close the issue.  I'll let that for you, if you want to do it.

The bottom line, of course, is that (octal-wise) AP11ROPE=Luminary99R1.

**Comment by thewonderidiot** [comment]
Let's close it! :smile: I'll open up some other issues related to the above.
